### PR TITLE
Fix BackupImpl holding stale references to db-components.

### PR DESCRIPTION
**Pull request by martinfurmanski**
The backup service will now resolve dependencies right before doing the backup.
Added a test for doing backups from any instance in HA.

### CODE REVIEW

File: enterprise/backup/src/main/java/org/neo4j/backup/BackupImpl.java [hunk 1]
```diff
@@ -37,29 +45,48 @@
...
-    public BackupImpl( StoreCopyServer storeCopyServer, Monitors monitors,
-                       LogicalTransactionStore logicalTransactionStore, TransactionIdStore transactionIdStore,
-                       LogFileInformation logFileInformation, Provider<StoreId> storeId )
+    public BackupImpl( GraphDatabaseAPI graphDatabaseAPI, Monitors monitors, Provider<StoreId> storeId )
     {
-        this.storeCopyServer = storeCopyServer;
-        this.logicalTransactionStore = logicalTransactionStore;
-        this.transactionIdStore = transactionIdStore;
-        this.logFileInformation = logFileInformation;
+        this.graphDatabaseAPI = graphDatabaseAPI;
```
Comment (davidegrohmann) [inline enterprise/backup/src/main/java/org/neo4j/backup/BackupImpl.java]: can't we inject the `DependencyResolver` instead?

**Comment by davidegrohmann** [comment]
@martinfurmanski isn't it possible to make the `BackupImpl` part of the life cycle management?

**Comment by martinfurmanski** [comment]
@davidegrohmann Yes, I got this comment before. I think no matter how you do it, this dependency management is messy.

**Comment by davidegrohmann** [comment]
@martinfurmanski You are right this way is no prettier, but hopefully it will make easier to refactor later on.

**Pull request merged**